src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -18,17 +18,1020 @@
+                                        sstables,
+                                        level + 1,
+                                        step * F,
+                                        maxSSTablesToCompact,
+                                        list);
+
+            int total = sstables.size();
+            // step defines the number of source sstables that are needed to reach this level (ignoring overwrites
+            // and deletions).
+            // For tiered compaction we will select batches of this many.
+            int pickSize = step;
+            if (W < 0)
+            {
+                // For levelled compaction all the sstables that would reach this level need to be compacted to one,
+                // so select the highest multiple of step that is available, but make sure we don't do a compaction
+                // bigger than the limit.
+                pickSize *= Math.min(total - pos, maxSSTablesToCompact) / pickSize;
+
+                if (pickSize == 0)  // Not enough sstables to reach this level, we can skip the processing below.
+                    return pos;     // Note: this cannot happen on the top level, but can on lower ones.
+            }
+
+            while (pos + pickSize <= total)
+            {
+                // Note that we assign these compactions to the level that would normally produce them, which means that
+                // they won't be taking up threads dedicated to the busy level.
+                // Normally sstables end up on a level when a compaction on the previous brings their size to the
+                // threshold (which corresponds to pickSize == step, always the case for tiered); in the case of
+                // levelled compaction, when we compact more than 1 but less than F sstables on a level (which
+                // corresponds to pickSize > step), it is an operation that is triggered on the same level.
+                list.add(CompactionPick.create(pickSize > step ? level : level - 1,
+                                               sstables.subList(pos, pos + pickSize)));
+                pos += pickSize;
+            }
+
+            // In the levelled case, if we had to adjust pickSize due to maxSSTablesToCompact, there may
+            // still be enough sstables to reach this level (e.g. if max was enough for 2*step, but we had 3*step).
+            if (pos + step <= total)
+            {
+                pickSize = ((total - pos) / step) * step;
+                list.add(CompactionPick.create(pickSize > step ? level : level - 1,
+                                               sstables.subList(pos, pos + pickSize)));
+                pos += pickSize;
+            }
+            return pos;
+        }
+
+        int WA()
+        {
+            return W >= 0 ? 1 : F; // W >= 0 => tiered compaction, <0 => leveled compaction
+        }
+
+        int RA()
+        {
+            return W >= 0 ? T : 1; // W >= 0 => tiered compaction, <0 => leveled compaction
```

Comment:
In theory this is just `T-1`.

What we can have in reality (for a shard count of 1 or unlimited number of threads), under full sustained capacity, is T tables from the level being compacted and the next T-1 already here (i.e. as soon as the compaction on the level finishes the next one is ready). So if we are looking for worst case under normal operation, this should be `T + T - 1` with an explanation why.

Reply:
Removed as these methods were unused.